healthcheck.go

```diff
@@ -53,6 +55,73 @@ type ClusterHealth struct {
 	Status string `json:"status"`
 }
 
+//indexcheck calls elasticsearch to check if an index from the client exists
+//What is a context?
+func (cli *Client) indexcheck(ctx context.Context) (code int, err error) {
+
+	clientElasticSearchIndices := cli.indices
+	/*
+		Is there a way write a test here maybe:
+		clientElasticSearchIndices := cli.indices
+	*/
+	logData := log.Data{"clientIndexResponse": cli.indices}
+	if err != nil {
+		log.Event(ctx, "failed to retrieve the indices from the client", log.ERROR, logData, log.Error(ErrorRetrievingClientIndices))
+		return 500, ErrorRetrievingClientIndices
+	}
+
+	for _, index := range clientElasticSearchIndices {
+
+		urlIndex := cli.url + "/" + index
+		logData := log.Data{"url": urlIndex}
+
+		URL, err := url.Parse(urlIndex)
+		if err != nil {
+			log.Event(ctx, "failed to create url for elasticsearch indexcheck", log.ERROR, logData, log.Error(err))
+			return 500, err
+		}
+
+		path := URL.String()
+		//What do the ["url"] brackets mean - is it resetting the value for the "url" key to the path
+		logData["url"] = path
+
+		//what is background context?
+		req, err := http.NewRequest("HEAD", path, nil)
+		if err != nil {
+			log.Event(ctx, "failed to create request for indexcheck call to elasticsearch", log.ERROR, logData, log.Error(err))
+			return 500, err
+		}
+
+		//Where do the credentials come from?
+		if cli.signRequests {
+			awsauth.Sign(req)
+		}
+
+		resp, err := cli.httpCli.Do(ctx, req)
+		if err != nil {
+			log.Event(ctx, "failed to call elasticsearch", log.ERROR, logData, log.Error(err))
+			return 500, err
+		}
+		defer resp.Body.Close()
+		logData["http_code for "+index] = resp.StatusCode
+
+		switch resp.StatusCode {
+
+		case 200:
+			log.Event(ctx, "An index exists", logData)
+		case 404:
+			log.Event(ctx, "404 status code returned in response", logData, log.ERROR, log.Error(ErrorIndexDoesNotExist))
```

Comment:
This error event could be more descriptive so that when we are debugging we can clearly see an error for "index does not exist" or whatever in the logs to help point us in the right direction.